### ISSUE TITLE
Issue reading large text values

### ISSUE THREAD

**Issue by ak-coram** [impossibl/pgjdbc-ng#489]
I've encountered a `TooLongFrameException` when reading larger (dozens
of MBs) text values and came up with a small Clojure example to
reproduce:

```clojure
{:deps {com.impossibl.pgjdbc-ng/pgjdbc-ng {:mvn/version "0.8.4"}
        org.postgresql/postgresql {:mvn/version "42.2.12"}
        seancorfield/next.jdbc {:mvn/version "1.0.445"}}}
```

```clojure
(let [ds (jdbc/get-datasource {:dbtype "pgsql" :dbname "maxtext" :port 5432})]
  (with-open [conn (jdbc/get-connection ds)]
    (jdbc/execute! conn [(str "CREATE TABLE IF NOT EXISTS t1 "
                              "(id SERIAL PRIMARY KEY, s TEXT DEFAULT '')")])
    (jdbc/execute! conn [(str "DELETE FROM t1")])
    (while true
      (jdbc/execute! conn [(str "INSERT INTO t1 (id, s) VALUES "
                                "(1, 'x') ON CONFLICT (id) DO UPDATE SET s = t1.s || t1.s")])
      (prn (count (:t1/s (first (jdbc/execute! conn [(str "SELECT s FROM t1")]))))))))
```

Running the above prints the following on my system:

```
1
2
4
8
16
32
64
128
256
512
1024
2048
4096
8192
16384
32768
65536
131072
262144
524288
1048576
2097152
4194304
8388608
16777216
Execution error (TooLongFrameException) at io.netty.handler.codec.LengthFieldBasedFrameDecoder/fail (LengthFieldBasedFrameDecoder.java:513).
Adjusted frame length exceeds 20971520: 33554443 - discarded
```

When switching to the `org.postgresql/postgresql` driver the same code
also fails, but only when trying to allocate a whole gigabyte:

```
1
2
4
8
16
32
64
128
256
512
1024
2048
4096
8192
16384
32768
65536
131072
262144
524288
1048576
2097152
4194304
8388608
16777216
33554432
67108864
134217728
268435456
536870912
Execution error (PSQLException) at org.postgresql.core.v3.QueryExecutorImpl/receiveErrorResponse (QueryExecutorImpl.java:2533).
오류: invalid memory alloc request size 1073741828
```

Is there something that could be done to increase the size of readable
text values when using pgjdbc-ng? Please let me know if you need
further information on this issue. Thanks!

Here's the full trace for the exception:

```
2. Unhandled java.sql.SQLException
   io.netty.handler.codec.TooLongFrameException: Adjusted frame length
   exceeds 20971520: 33554443 - discarded

          PGStatement.java:  384  com.impossibl.postgres.jdbc.PGStatement/executeStatement
  PGPreparedStatement.java:  339  com.impossibl.postgres.jdbc.PGPreparedStatement/execute
            result_set.clj:  559  next.jdbc.result-set/stmt->result-set
            result_set.clj:  554  next.jdbc.result-set/stmt->result-set
            result_set.clj:  645  next.jdbc.result-set/eval15322/fn
             protocols.clj:   33  next.jdbc.protocols/eval14584/fn/G
                  jdbc.clj:  220  next.jdbc/execute!
                  jdbc.clj:  210  next.jdbc/execute!
                      REPL:   12  core/eval15599
                      REPL:    4  core/eval15599
             Compiler.java: 7177  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
    interruptible_eval.clj:   91  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  665  clojure.core/apply
                  core.clj:  660  clojure.core/apply
                regrow.clj:   20  refactor-nrepl.ns.slam.hound.regrow/wrap-clojure-repl/fn
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  155  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  190  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  189  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  834  java.lang.Thread/run

1. Caused by io.netty.handler.codec.TooLongFrameException
   Adjusted frame length exceeds 20971520: 33554443 - discarded

LengthFieldBasedFrameDecoder.java:  513  io.netty.handler.codec.LengthFieldBasedFrameDecoder/fail
LengthFieldBasedFrameDecoder.java:  491  io.netty.handler.codec.LengthFieldBasedFrameDecoder/failIfNecessary
LengthFieldBasedFrameDecoder.java:  378  io.netty.handler.codec.LengthFieldBasedFrameDecoder/exceededFrameLength
LengthFieldBasedFrameDecoder.java:  421  io.netty.handler.codec.LengthFieldBasedFrameDecoder/decode
LengthFieldBasedFrameDecoder.java:  334  io.netty.handler.codec.LengthFieldBasedFrameDecoder/decode
 ByteToMessageDecoder.java:  505  io.netty.handler.codec.ByteToMessageDecoder/decodeRemovalReentryProtection
 ByteToMessageDecoder.java:  444  io.netty.handler.codec.ByteToMessageDecoder/callDecode
 ByteToMessageDecoder.java:  283  io.netty.handler.codec.ByteToMessageDecoder/channelRead
AbstractChannelHandlerContext.java:  374  io.netty.channel.AbstractChannelHandlerContext/invokeChannelRead
AbstractChannelHandlerContext.java:  360  io.netty.channel.AbstractChannelHandlerContext/invokeChannelRead
AbstractChannelHandlerContext.java:  352  io.netty.channel.AbstractChannelHandlerContext/fireChannelRead
DefaultChannelPipeline.java: 1422  io.netty.channel.DefaultChannelPipeline$HeadContext/channelRead
AbstractChannelHandlerContext.java:  374  io.netty.channel.AbstractChannelHandlerContext/invokeChannelRead
AbstractChannelHandlerContext.java:  360  io.netty.channel.AbstractChannelHandlerContext/invokeChannelRead
DefaultChannelPipeline.java:  931  io.netty.channel.DefaultChannelPipeline/fireChannelRead
AbstractNioByteChannel.java:  163  io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe/read
         NioEventLoop.java:  700  io.netty.channel.nio.NioEventLoop/processSelectedKey
         NioEventLoop.java:  635  io.netty.channel.nio.NioEventLoop/processSelectedKeysOptimized
         NioEventLoop.java:  552  io.netty.channel.nio.NioEventLoop/processSelectedKeys
         NioEventLoop.java:  514  io.netty.channel.nio.NioEventLoop/run
SingleThreadEventExecutor.java: 1044  io.netty.util.concurrent.SingleThreadEventExecutor$6/run
    ThreadExecutorMap.java:   74  io.netty.util.internal.ThreadExecutorMap$2/run
               Thread.java:  834  java.lang.Thread/run
```

**Comment by kdubb** [comment]
See the "Field Length Max" setting in the user guide.

http://impossibl.github.io/pgjdbc-ng/docs/0.8.4/user-guide/#_system_settings

**Comment by kdubb** [comment]
FYI the default is 20mb as you've experienced

**Comment by ak-coram** [comment]
Thank you for your quick reply, I definitely should have found that.

I'm still having some trouble using it though. Should this be in bytes? Whatever I try setting it to, I always get the following exception `(System/setProperty "pgjdbc.field.length.max" "123")`.

```
1. Unhandled java.lang.StringIndexOutOfBoundsException
   begin 0, end 123, length 11

               String.java: 3319  java.lang.String/checkBoundsBeginEnd
               String.java: 1874  java.lang.String/substring
               String.java: 1913  java.lang.String/subSequence
              Strings.java:  240  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decodeNativeValue
              Strings.java:  221  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decodeNativeValue
         BaseDecoders.java:  277  com.impossibl.postgres.system.procs.AutoConvertingTextDecoder/decodeValue
         BaseDecoders.java:  216  com.impossibl.postgres.system.procs.BaseTextDecoder/decode
              Strings.java:  221  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decode
         BaseDecoders.java:  194  com.impossibl.postgres.system.procs.BaseTextDecoder/decode
        BufferRowData.java:  145  com.impossibl.postgres.protocol.v30.BufferRowData/getField
         BasicContext.java:  392  com.impossibl.postgres.system.BasicContext/loadServerLocales
         BasicContext.java:  382  com.impossibl.postgres.system.BasicContext/init
   PGDirectConnection.java:  288  com.impossibl.postgres.jdbc.PGDirectConnection/init
       ConnectionUtil.java:  319  com.impossibl.postgres.jdbc.ConnectionUtil/createConnection
       ConnectionUtil.java:  279  com.impossibl.postgres.jdbc.ConnectionUtil/createConnection
             PGDriver.java:  119  com.impossibl.postgres.jdbc.PGDriver/connect
        DriverManager.java:  677  java.sql.DriverManager/getConnection
        DriverManager.java:  189  java.sql.DriverManager/getConnection
            connection.clj:  137  next.jdbc.connection/get-driver-connection
            connection.clj:  132  next.jdbc.connection/get-driver-connection
            connection.clj:  321  next.jdbc.connection/url+etc/reify
            connection.clj:  343  next.jdbc.connection/make-connection
            connection.clj:  331  next.jdbc.connection/make-connection
            connection.clj:  366  next.jdbc.connection/eval13059/fn
             protocols.clj:   24  next.jdbc.protocols/eval12823/fn/G
                  jdbc.clj:  149  next.jdbc/get-connection
                  jdbc.clj:  132  next.jdbc/get-connection
                      REPL:    9  core/eval15994
                      REPL:    6  core/eval15994
             Compiler.java: 7177  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
    interruptible_eval.clj:   91  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  665  clojure.core/apply
                  core.clj:  660  clojure.core/apply
                regrow.clj:   20  refactor-nrepl.ns.slam.hound.regrow/wrap-clojure-repl/fn
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  155  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  190  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  189  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  834  java.lang.Thread/run
```

**Comment by kdubb** [comment]
How are you setting the value exactly?

Also you can also try using `Statement.setMaxFieldSize` sets the same value via the JDBC interface and isn't a system wide change.

**Comment by ak-coram** [comment]
I've used `java.lang.System.setProperty` to set the value. I've now also tried setting it only on the affected statement:

```clojure
(ns core
  (:require [next.jdbc :as jdbc]
            [next.jdbc.prepare :as p]))

(let [ds (jdbc/get-datasource {:dbtype "pgsql"
                               :dbname "maxtext"
                               :port 5432})]
  (with-open [conn (jdbc/get-connection ds)]
    (jdbc/execute! conn [(str "CREATE TABLE IF NOT EXISTS t1 "
                              "(id SERIAL PRIMARY KEY, s TEXT DEFAULT '')")])
    (jdbc/execute! conn [(str "DELETE FROM t1")])
    (let [stmt (jdbc/prepare conn ["SELECT s FROM t1"])]
      (.setMaxFieldSize stmt (* 1024 1024 100))
      (while true
        (jdbc/execute! conn [(str "INSERT INTO t1 (id, s) VALUES "
                                  "(1, 'x') ON CONFLICT (id) DO UPDATE SET s = t1.s || t1.s")])
        (prn (count (:t1/s (jdbc/execute-one! (p/set-parameters stmt [])))))))))
```

The result is a similar error:

```
1. Unhandled java.lang.StringIndexOutOfBoundsException
   begin 0, end 104857600, length 1

               String.java: 3319  java.lang.String/checkBoundsBeginEnd
               String.java: 1874  java.lang.String/substring
               String.java: 1913  java.lang.String/subSequence
              Strings.java:  240  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decodeNativeValue
              Strings.java:  221  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decodeNativeValue
         BaseDecoders.java:  277  com.impossibl.postgres.system.procs.AutoConvertingTextDecoder/decodeValue
         BaseDecoders.java:  216  com.impossibl.postgres.system.procs.BaseTextDecoder/decode
              Strings.java:  221  com.impossibl.postgres.system.procs.Strings$TxtDecoder/decode
         BaseDecoders.java:  194  com.impossibl.postgres.system.procs.BaseTextDecoder/decode
        BufferRowData.java:  145  com.impossibl.postgres.protocol.v30.BufferRowData/getField
          PGResultSet.java: 1858  com.impossibl.postgres.jdbc.ListScroller/getRowField
          PGResultSet.java:  301  com.impossibl.postgres.jdbc.PGResultSet/getObj
          PGResultSet.java:  870  com.impossibl.postgres.jdbc.PGResultSet/getObject
            result_set.clj:  152  next.jdbc.result-set.MapResultSetBuilder/with_column
            result_set.clj:  381  next.jdbc.result-set/row-builder/fn
            LongRange.java:  233  clojure.lang.LongRange/reduce
                  core.clj: 6827  clojure.core/reduce
                  core.clj: 6810  clojure.core/reduce
            result_set.clj:  381  next.jdbc.result-set/row-builder
            result_set.clj:  377  next.jdbc.result-set/row-builder
            result_set.clj:  696  next.jdbc.result-set/eval13617/fn
             protocols.clj:   33  next.jdbc.protocols/eval12855/fn/G
                  jdbc.clj:  237  next.jdbc/execute-one!
                  jdbc.clj:  226  next.jdbc/execute-one!
                      REPL:   17  core/eval13737
                      REPL:    5  core/eval13737
             Compiler.java: 7177  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
    interruptible_eval.clj:   91  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java:  137  clojure.lang.RestFn/applyTo
                  core.clj:  665  clojure.core/apply
                  core.clj:  660  clojure.core/apply
                regrow.clj:   20  refactor-nrepl.ns.slam.hound.regrow/wrap-clojure-repl/fn
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  155  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  190  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  189  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  834  java.lang.Thread/run
```

**Comment by kdubb** [comment]
@ak-coram There is a bug when setting it explicitly.  I'm building a PR right now. ☹️

**Comment by ak-coram** [comment]
@kdubb: I'm glad that at least creating this issue wasn't completely useless in the end :)

Thanks again for your help! I'll test the changes as soon as they're up.

**Comment by kdubb** [comment]
In my haste I accidentally pushed directly to `develop`.  So, providing the CI completes successfully it'll be in `0.9-SNAPSHOT`

**Comment by kdubb** [comment]
@ak-coram The CI is passed and fix should be in `0.9-SNAPSHOT`!

**Comment by ak-coram** [comment]
I can now set the maximum field size successfully (via both `System.setProperty` and directly on the statement with `.setMaxFieldSize`). When I use small values for the field size, the results get truncated to that length. However it still fails with the same exception when the field is larger than 20MB:

````
Execution error (TooLongFrameException) at io.netty.handler.codec.LengthFieldBasedFrameDecoder/fail (LengthFieldBasedFrameDecoder.java:513).
Adjusted frame length exceeds 20971520: 33554443 - discarded
````

Increasing the maximum field size above the default seems to have no effect either.

Here's the full test:

```clojure
(ns core
  (:require [next.jdbc :as jdbc]
            [next.jdbc.prepare :as p]))

(let [ds (jdbc/get-datasource {:dbtype "pgsql"
                               :dbname "maxtext"
                               :port 5432})]
  (with-open [conn (jdbc/get-connection ds)]
    (jdbc/execute! conn [(str "CREATE TABLE IF NOT EXISTS t1 "
                              "(id SERIAL PRIMARY KEY, s TEXT DEFAULT '')")])
    (jdbc/execute! conn [(str "DELETE FROM t1")])
    (let [stmt (jdbc/prepare conn ["SELECT s FROM t1"])]
      (.setMaxFieldSize stmt 200)
      (while true
        (jdbc/execute! conn [(str "INSERT INTO t1 (id, s) VALUES "
                                  "(1, 'x') ON CONFLICT (id) DO UPDATE SET s = t1.s || t1.s")])
        (prn (count (:t1/s (jdbc/execute-one! (p/set-parameters stmt [])))))))))
```

And the output:

```
1
2
4
8
16
32
64
128
200
200
200
200
200
200
200
200
200
200
200
200
200
200
200
200
200
Execution error (TooLongFrameException) at io.netty.handler.codec.LengthFieldBasedFrameDecoder/fail (LengthFieldBasedFrameDecoder.java:513).
Adjusted frame length exceeds 20971520: 33554443 - discarded
```

Any ideas? Thanks!

**Comment by kdubb** [comment]
@ak-coram Apologies, I actually sent you to the incorrect setting.  There is a protocol specific hard limit:

`Protocol Message Size Max` in
http://impossibl.github.io/pgjdbc-ng/docs/0.8.4/user-guide/#_protocol_settings

You'll notice it even has the default listed of 20mb 🤦‍♂️

The reason there are essentially two settings is because field `field.length.max` can be set in multiple ways (e.g. system settings and the JDBC API) and can be changed dynamically. Alternatively, `protocol.message.size.max` is static and must be set before connection time.

**Comment by ak-coram** [comment]
@kdubb: I've tested `pgjdbc.protocol.message.size.max` it and it works great even in version 0.8.4. Thanks again for your immediate help. At least we've discovered this other bug which got resolved in the end :)

Not sure how often this comes up, but perhaps the docs could mention the kind of exception that is thrown when the field size exceeds the hard limit (or the exception could be wrapped into something which mentions the setting). Since I don't tend to use Netty directly myself, the errors seem a little bit cryptic without further context.
